### PR TITLE
[SHLWAPI][SHLWAPI_APITEST][SDK] SHCreatePropertyBagOnRegKey

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Follow-up to #5511. Implementing missing features...
JIRA issue: [CORE-9283](https://jira.reactos.org/browse/CORE-9283)

## Proposed changes

- Add `CRegPropertyBag` class.
- Implement `SHCreatePropertyBagOnRegKey` function by using `CRegPropertyBag`.
- Strengthen `SHPropertyBag` testcase in `shlwapi_apitest`.

## TODO

- [x] Do small tests.

## Comparison

WinXP:
WinXP
Successful.

Win2k3:
Win2k3
Successful.

Win10:
Win10
Successful.

ReactOS (AFTER):
after
Successful.

### CODE REVIEW

File: dll/win32/shlwapi/propbag.cpp [hunk 1]
```diff
@@ -251,3 +251,323 @@ SHCreatePropertyBagOnMemory(_In_ DWORD dwMode, _In_ REFIID riid, _Out_ void **pp
...
+        nAccess |= KEY_WRITE;
+
+    LONG error;
+    if (m_dwMode & STGM_CREATE)
+        error = ::RegCreateKeyExW(hKey, lpSubKey, 0, NULL, 0, nAccess, NULL, &m_hKey, NULL);
+    else
+        error = ::RegOpenKeyExW(hKey, lpSubKey, 0, nAccess, &m_hKey);
+
+    if (error != ERROR_SUCCESS)
+        return (LOWORD(error) | 0x80070000); // FIXME: What is 0x80070000?
```
Comment (HBelusca) [inline dll/win32/shlwapi/propbag.cpp:303]: Do you get the 8070000 in WinXP or 2k3 ? Or in Win10 ?

File: modules/rostests/apitests/shlwapi/SHPropertyBag.cpp [hunk 2]
```diff
@@ -456,9 +456,129 @@ static void SHPropertyBag_OnMemory(void)
...
+    }
+
+    IPropertyBag *pPropBag;
+    HRESULT hr;
+
+    // Try to create new registry key
+    RegDeleteKeyW(hKey, L"PropBagTest");
+    hr = SHCreatePropertyBagOnRegKey(hKey, L"PropBagTest", 0,
+                                     IID_IPropertyBag, (void **)&pPropBag);
+    ok_long(hr, 0x80070002);
```
Comment (HBelusca) [inline modules/rostests/apitests/shlwapi/SHPropertyBag.cpp:481]:
```suggestion
    ok_long(hr, HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND));
```

File: modules/rostests/apitests/shlwapi/SHPropertyBag.cpp [hunk 3]
```diff
@@ -456,9 +456,129 @@ static void SHPropertyBag_OnMemory(void)
...
+
+    // Try to create new registry key
+    RegDeleteKeyW(hKey, L"PropBagTest");
+    hr = SHCreatePropertyBagOnRegKey(hKey, L"PropBagTest", STGM_READWRITE,
+                                     IID_IPropertyBag, (void **)&pPropBag);
+    ok_long(hr, 0x80070002);
```
Comment (HBelusca) [inline modules/rostests/apitests/shlwapi/SHPropertyBag.cpp:487]:
```suggestion
    ok_long(hr, HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND));
```

File: dll/win32/shlwapi/propbag.cpp [hunk 4]
```diff
@@ -251,3 +251,323 @@ SHCreatePropertyBagOnMemory(_In_ DWORD dwMode, _In_ REFIID riid, _Out_ void **pp
...
+
+        case VT_BOOL:
+        case VT_I1:
+        case VT_I2:
+        case VT_I4:
+        case VT_UI1:
+        case VT_UI2:
+        case VT_UI4:
+        case VT_INT:
+        case VT_UINT:
```
Comment (HBelusca) [inline dll/win32/shlwapi/propbag.cpp:489]:
```suggestion
        case VT_UINT:
        {
```

File: dll/win32/shlwapi/propbag.cpp [hunk 5]
```diff
@@ -251,3 +251,323 @@ SHCreatePropertyBagOnMemory(_In_ DWORD dwMode, _In_ REFIID riid, _Out_ void **pp
...
+            hr = ::VariantChangeType(&vargTemp, pvari, 0, VT_UI4);
+            if (FAILED(hr))
+                return hr;
+
+            error = SHSetValueW(m_hKey, NULL, pszPropName, REG_DWORD, &V_UI4(&vargTemp), sizeof(DWORD));
+            if (error)
+                hr = E_FAIL;
+
+            ::VariantClear(&vargTemp);
+            break;
```
Comment (HBelusca) [inline dll/win32/shlwapi/propbag.cpp:499]:
```suggestion
            break;
        }
```

File: sdk/include/reactos/shlwapi_undoc.h [hunk 6]
```diff
@@ -91,6 +91,7 @@ BOOL WINAPI SHExpandEnvironmentStringsForUserW(HANDLE, LPCWSTR, LPWSTR, DWORD);
 
 
 BOOL WINAPI SHIsEmptyStream(IStream*);
+HRESULT WINAPI IStream_Size(IStream *lpStream, ULARGE_INTEGER* lpulSize);
```
Comment (HBelusca) [inline sdk/include/reactos/shlwapi_undoc.h:94]: I will add the missing IStream_* functions in shlwapi.h in a separate PR.

File: dll/win32/shlwapi/propbag.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shlwapi/propbag.cpp]: Got in WinXP, Win2k3, and Win10.

**Review by JoachimHenze — APPROVED** [review]

File: sdk/include/reactos/shlwapi_undoc.h (hunk 6 above)
Comment (katahiromz) [reply sdk/include/reactos/shlwapi_undoc.h:94]: Now or future?

Comment (HBelusca) [reply sdk/include/reactos/shlwapi_undoc.h:94]: later today.

**Pull request merged**